### PR TITLE
This commit carries rails assignment with recommended changes

### PR DESCRIPTION
**Pull request by Suresh-Yenni**
Please review my Rails and ReactJS assignment

### CODE REVIEW

File: rails_assignments/colleges_students_react/app/assets/javascripts/components/_student.js.jsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  getInitialState : function() {
+      return {
+        student : {},
+        marks_change: {},
+        total_marks: {},
+        display:[],
+        index: -1,
+        group_on : ""
+      };
+
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/assets/javascripts/components/_student.js.jsx]: remove line

File: rails_assignments/colleges_students_react/app/controllers/colleges_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+class CollegesController < ApplicationController
+
+	def home; 
+	end
+
+	def search
+		str = params[:college][:name]
+		@college = College.where('name like ?', "%#{str}%")
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/controllers/colleges_controller.rb]: Use select here

File: rails_assignments/colleges_students_react/app/views/colleges/search.html.erb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        .jump:hover
+        {
+           font-size:120%;
+        }
+    </style>
+    <h1 align="center" style="color: #0099ff;"><%= t("match_available") %></h1>
+  </head>
+  <body bgcolor="#f1f1f1">
+    <div class="block">
+      <% if !@college.empty? %><br><br>
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/views/colleges/search.html.erb]: use blank wherever necessary

File: rails_assignments/colleges_students_react/app/views/students/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+          </tr>
+          <% @details.each do |stu, clg| %>
+           <tr>
+             <td><%= stu.regid %></td>
+             <td><%= stu.dept %></td>
+             <td><%= stu.maths %></td>
+             <td><%= stu.physics %></td>
+             <td><%= stu.chemistry %></td>
+             <td><%= stu.year %></td>
+             <th><%= clg %></td>
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/views/students/index.html.erb]: use back ref here stu.college.name

File: rails_assignments/colleges_students_react/app/controllers/students_controller.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+		render json: [@students, @marks_change, @total_marks, @display]
+	end
+
+	def query; end
+
+	def index
+		@students = Student.all
+		colleges = []
+		@students.each { |student| colleges.push(student.college.name) }
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/controllers/students_controller.rb]: Use back referencing and reducing the usage of variables here

File: rails_assignments/colleges_students_react/app/controllers/students_controller.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+	def new
+		@student = Student.new
+	end
+
+	def create
+		student = Student.new(params.require(:student).permit!)
+		student.save	
+		col_name = College.select('name').find(params[:student][:college_id])
+		col_name = col_name.name[0...2]
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/controllers/students_controller.rb]: merge the two lines into single line col_name

File: rails_assignments/colleges_students_react/app/models/student.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    def self.compute_change_and_total!(marks_change, total_marks, student, year1, year2, total)
+        student.keys.each do |key|
+            change_row = {"maths"=> 0, "physics" => 0, "chemistry" => 0}
+            total_row = {"maths"=> 0, "physics" => 0, "chemistry" => 0}
+            math1, math2, phy1, phy2, chem1, chem2 = [0,0,0,0,0,0]
+            student[key].each do |stu|
+                total_row["maths"] += stu['maths']
+                total_row["physics"] += stu['physics']
+                total_row["chemistry"] += stu['chemistry']
+                if !year1.nil? && !year2.nil?
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/models/student.rb]: use unless

File: rails_assignments/colleges_students_react/app/models/student.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                        phy1 = stu['physics']
+                        chem1 = stu['chemistry']
+                    elsif stu['year'] == year2
+                        math2 = stu['maths']
+                        phy2 = stu['physics']
+                        chem2 = stu['chemistry']
+                    end
+                end
+            end
+            if !year1.nil? && !year2.nil?
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/models/student.rb]: use unless

File: rails_assignments/colleges_students_react/app/views/colleges/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        .jump:hover
+        {
+            font-size:120%;
+        }
+    </style>
+    <h1 align="center" style="color: #0099ff;"><%= t("available_colleges") %></h1>
+  </head>
+  <body bgcolor="#f1f1f1">
+    <div class="block">
+      <% if !@college.empty? %><br><br>
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/views/colleges/index.html.erb]: use unless

File: rails_assignments/colleges_students_react/app/views/colleges/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  }
+  .jump:hover
+  {
+    font-size:120%;
+  }
+</style>
+<h1 align="center" style="color: #0099ff;"><%= t("college_students") %></h1>
+<body bgcolor="#f1f1f1">
+  <div class="block">
+     <% if !@student.empty? %><br><br>
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/views/colleges/show.html.erb]: use unless

File: rails_assignments/colleges_students_react/app/views/students/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      .jump:hover
+      {
+         font-size:120%;
+      }
+    </style>
+    <h1 align="center" style="color: #0099ff;"><%= t("available_students") %></h1>
+  </head>
+  <body bgcolor="#f1f1f1">
+    <div class="block">
+      <% if !@students.blank? %><br><br>
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/views/students/index.html.erb]: use unless

File: rails_assignments/ecommerce/app/controllers/customers_controller.rb [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def new
+    @customers = Customer.new
+  end
+
+  def show
+  	customer = Customer.find(params[:id])
+  	customer_orders = customer.orders
+    items = []
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/controllers/customers_controller.rb]: indentation

File: rails_assignments/ecommerce/app/controllers/customers_controller.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    @customers = Customer.all
+  end
+
+  def new
+    @customers = Customer.new
+  end
+
+  def show
+  	customer = Customer.find(params[:id])
+  	customer_orders = customer.orders
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/controllers/customers_controller.rb]: Reduce it to single query Customer.find().orders

File: rails_assignments/ecommerce/app/views/customers/show.html.erb [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+</style>
+<h1 align="center" style="color: #0099ff;">Showing Customer Orders</h1>
+<body bgcolor="#f1f1f1">
+  <div class="block">
+     <% if !@details.empty? %><br><br>
+       <table border="1">
+            <% @details.each do |ord, item| %>
+              <tr>
+                <td><%= ord.date %></td>
+                <td><%= item.item_name %></td>
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/views/customers/show.html.erb]: Use backreferencing ord.item.item_name, make necessary changes in customers_controller.rb

File: rails_assignments/ecommerce/app/controllers/orders_controller.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def index
+    @details = []
+    Order.all.each { |odr| @details.push([odr.customer.name, odr.item.item_name]) }
+  end
+
+  def new
+    @orders = Order.new
+    @customers = Customer.all
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/controllers/orders_controller.rb]: use select here

File: rails_assignments/ecommerce/app/views/orders/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+       color:#0099ff;
+  }
+</style>
+<h1 align="center" style="color: #0099ff;">Making New Order</h1>
+<body bgcolor="#f1f1f1">
+  <div class="block"><br><br>
+    <%= form_for :orders, url: orders_path do |f|%>
+       <p>
+         <%= f.collection_select(:customer_id, Customer.all, :id, :name) %><br><br>
+         <%= f.collection_select(:item_id, Item.all, :id, :item_name) %><br><br>
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/views/orders/new.html.erb]: See if you can reduce the above two queries

File: rails_assignments/ecommerce/app/controllers/orders_controller.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
+class OrdersController < ApplicationController
+
+  def home
+  end
+
+  def index
+    @details = []
+    Order.all.each { |odr| @details.push([odr.customer.name, odr.item.item_name]) }
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/controllers/orders_controller.rb]: use includes, select here

File: rails_assignments/ecommerce/app/views/customers/index.html.erb [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        .jump:hover
+        {
+           font-size:120%;
+        }
+    </style>
+    <h1 align="center" style="color: #0099ff;">All Available Customers</h1>
+  </head>
+  <body bgcolor="#f1f1f1">
+    <div class="block">
+      <% if !@customers.empty? %><br><br>
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/views/customers/index.html.erb]: use blank

File: rails_assignments/colleges_students_react/app/assets/javascripts/components/_student.js.jsx [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+          error: function(xhr, textStatus, errorThrown){
+                     alert("Error Occurred: No search done");
+                  }
+      });
+  },
+  
+  render(){
+   
+      return( <div>
+                <p> GROUP BY </p>
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/assets/javascripts/components/_student.js.jsx]: Convert these into smaller class objects of react like header, etc.. ref: https://www.fullstackreact.com/30-days-of-react/day-5/

File: rails_assignments/colleges_students_react/app/controllers/colleges_controller.rb [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+		@college = College.college_on_id(params[:id])
+	end
+
+	def update
+		College.college_on_id(params[:id]).update_attributes(params.require(:college).permit!)
+		redirect_to colleges_path
+	end
+
+	def destroy
+		College.college_on_id(params[:id]).destroy
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/controllers/colleges_controller.rb]: use find_by_id

File: rails_assignments/colleges_students_react/app/controllers/students_controller.rb [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+		@students = Student.order(params[:sort])
+		@students = @students.all.group_by(&:"#{params[:group]}")
+		@display = params[:display]
+
+		first = 0
+		second = 0
+		@marks_change = {}
+		@total_marks = {}
+
+		year1 = !params[:year1].empty? ? Integer(params[:year1]) : nil
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/controllers/students_controller.rb]: you can use params[x].to_i here. Let me know if there is any performance reason

File: rails_assignments/ecommerce/app/controllers/orders_controller.rb [hunk 22]
```diff
@@ -0,0 +1,27 @@
+class OrdersController < ApplicationController
+
+  def home
+  end
+
+  def index
+    @orders = Order.includes(:customer, :item)
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/controllers/orders_controller.rb]: Order.includes(:customer, :item).pluck("customers.name, items.item_name")

File: rails_assignments/ecommerce/app/models/customer.rb [hunk 23]
```diff
@@ -0,0 +1,10 @@
+class Customer < ApplicationRecord
+
+  validates :city, :name, presence: true, format: { with: /\A[a-zA-Z]+\z/ }
+  validates :email, presence: true
+  validates :pincode, presence: true, numericality: { only_integer: true}, format: { with: /\A[+-]?\d+\z/ }
+  has_many :orders, dependent: :destroy
+
+  scope :cutomer_on_id, -> (id)  { Customer.find(id)}
```
Comment (prashanth-rg) [inline rails_assignments/ecommerce/app/models/customer.rb]: use find_by_id instead

File: rails_assignments/colleges_students_react/app/controllers/colleges_controller.rb [hunk 24]
```diff
@@ -0,0 +1,41 @@
+class CollegesController < ApplicationController
+
+	def home;
+	end
+
+	def search
+		@college = College.select('id, name, establish').where('name like ?', "%#{params[:college][:name]}%")
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/controllers/colleges_controller.rb]: see if you can use College model scope here

File: rails_assignments/colleges_students_react/app/views/colleges/index.html.erb [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+        .jump:hover
+        {
+            font-size:120%;
+        }
+    </style>
+    <h1 align="center" style="color: #0099ff;"><%= t("available_colleges") %></h1>
+  </head>
+  <body bgcolor="#f1f1f1">
+    <div class="block">
+      <% unless @college.empty? %><br><br>
```
Comment (prashanth-rg) [inline rails_assignments/colleges_students_react/app/views/colleges/index.html.erb]: use blank? wherever necessary

File: rails_assignments/ecommerce/app/models/customer.rb (hunk 23 above)
Comment (Suresh-Yenni) [inline rails_assignments/ecommerce/app/models/customer.rb]: I used Scope here to just learn the Scope concept

**Pull request merged**